---
title: "DESeq2 Analysis"
author: '`r paste0("BBC, Analyst: ", stringr::str_to_title(stringr::str_replace_all(Sys.getenv("USER"), "\\.", " ")  ))`'
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    theme: yeti
    code_folding: hide
    self_contained: yes
    toc: true
    toc_depth: 5
    toc_float:
      collapsed: false
      smooth_scroll: false
    number_sections: true
params:
  se_obj: ""
  comparison_name: ""
  group_test: ""
  group_reference: ""
  fdr_cutoff: ""
  genes_of_interest: ""
---

# Analysis

```{r keep_figures, cache=TRUE}
# this chunk is just to keep the _files directory even when we turn off cacheing
```

```{r starttime, echo=TRUE, warning=TRUE, message=TRUE, cache=FALSE, cache.lazy = FALSE}
# save start time for script
start_tm <- Sys.time()
# start_tm
```

```{r make_outdir, echo=TRUE, warning=TRUE, message=TRUE, cache=FALSE, cache.lazy = FALSE}
outdir <- file.path("deseq2_out_files", params$comparison_name)

dir.create(outdir, recursive=TRUE, showWarnings = FALSE)
if (!dir.exists(outdir)) {
  stop(paste0("Failed to create the output directory: '", outdir, "'. ",
              "Please check the 'comparison_name' parameter in the config file, ensure the path is valid, ",
              "and verify that you have sufficient permissions to create directories."))
}
# message("Output directory: ", outdir)
```


This analysis was started on `r format(start_tm, "%B %d, %Y at %H:%M")`. In addition to this full report, individual result files can be accessed in the "DE Results" tab, under the "Supplementary files" header.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, warning=TRUE, message=TRUE, cache=FALSE, cache.lazy = FALSE, dev=c('png','pdf'), 
                      fig.width=4, fig.height=4, fig.path=paste0(outdir, "/individual_figures/"))
```

## Data Processing

### Set Up DESeq2 Analysis

```{r load_pkges}
suppressPackageStartupMessages({
  library(dplyr)
  library(stringr)
  library(ggplot2)
  library(readr)
  library(ggrepel)
  library(ComplexHeatmap)
  library(DESeq2)
  library(patchwork)
  library(vegan)
})
```

<!-- Set up your DESeq object -->

```{r read_se}
se <- readRDS(paste0("../../", params$se_obj))

# subset se to just samples in this comparison (relevant esp if variance is diff between groups)
se <- se[, colData(se)$group %in% c(params$group_test, params$group_reference)]

# factor the group column to make sure the reference group is the first level
se$group <- factor(se$group, levels=c(params$group_reference, params$group_test))
```


```{r check_all_assays, eval=FALSE}
# Let's take a look to see what assays are stored in the SummarizedExperiment object. 
assayNames(se)
```


```{r check_assays}
# Note that DESeq2 assumes the first assay is the raw counts.
stopifnot(assayNames(se)[1] == "counts")
```


```{r look_se, eval=FALSE}
# To print more information about this SummarizedExperiment object, you can just type its name. 
se
```

Set up analysis using counts data from the previous steps of the workflow, and 
basic meta data given from units.tsv, with group as the variable we're testing.
**Note that no covariates are included here**; if you have a more complex 
experimental setup with additional covariates and/or confounders, 
this model is overly simplistic and won't accurately model your data --
please modify and/or contact the BBC.

<!-- The counts and the meta data need to be stored inside a special DESeq2 object called a 'DESeqDataSet'.  -->
<!-- Here, we also specify that each gene will be fit with a model design of '~ group'. -->

```{r make_dds, message=FALSE}
dds <- DESeqDataSet(se, design = ~ group)
```

### Remove genes with low/no expression

We cannot do meaningful analyses of genes with very low counts. Keeping only genes
with 10 or more total read counts across samples will speed up the analysis.

```{r filter_low_genes}
# prefilter genes, keeping only genes with 10 or more total read counts across samples
total_genes <- nrow(dds)
# message(str_glue("Total number of genes: {total_genes}"))

keep <- rowSums(counts(dds)) >= 10
# message(str_glue("Keeping {sum(keep)} genes."))

dds <- dds[keep, ]
```

The original number of genes was `r total_genes`, and the number of genes kept after filtering is `r sum(keep)`.

### Different normalization approaches for different biases

[Types of biases in RNA-seq](https://vanandelinstitute-my.sharepoint.com/:b:/g/personal/kin_lau_vai_org/EcruvwL-OrBIvCzXZ7HMPlcBo65fu0pucrivMmCwzM98dA?e=yCkfTa)


### Run the DE workflow

The [DESeq](https://www.rdocumentation.org/packages/DESeq2/versions/1.12.3/topics/DESeq) function is a convenience function from DESeq2 that estimates size factors (normalization) and fits negative binomial GLMs.
This includes the following steps:  

- estimating size factors  
- estimating dispersions  
- gene-wise dispersion estimates  
- mean-dispersion relationship  
- final dispersion estimates  
- fitting model and testing  

```{r run_deseq2}
dds <- DESeq(dds, quiet = TRUE)
# message(paste0("Coefficient names are: ", paste(resultsNames(dds), collapse = " ")))
```

Coefficients in the model are: `r paste(resultsNames(dds), collapse = ", ")`.

After the models are fitted, we can test specific pairs of groups for differential expression. For DESeq2, it is recommended to provide the significance cutoff that you wish to use as it affects the multiple testing correction procedure (see [docs](https://www.rdocumentation.org/packages/DESeq2/versions/1.12.3/topics/results)). Here, we'll use the false discovery rate (FDR) cutoff that was provided in the config file (`r params$fdr_cutoff`).

```{r run_contrast}
contrast <- c("group", params$group_test, params$group_reference)
fdr_cutoff <- params$fdr_cutoff

res <- results(dds, contrast=contrast, alpha=fdr_cutoff)
res <- res[order(res$pvalue), ]
```

### Summarize DE results

Below is a summary of the DE results, including the number of upregulated and downregulated genes, and the total number of genes tested.

```{r de_summ}
df <- as.data.frame(res)
data.frame(
  Up=sum(df$padj <= fdr_cutoff & df$log2FoldChange > 0, na.rm = TRUE),
  Down=sum(df$padj <= fdr_cutoff & df$log2FoldChange < 0, na.rm = TRUE),
  Tested=sum(!is.na(df$padj))
) %>%
  knitr::kable() %>%
  kableExtra::kable_styling(full_width = FALSE, font_size = 14) 
```


### Shrink log fold changes for lowly expressed genes

This step does not affect the identification of DE genes, but it can be useful to perform this to obtain more reliable estimates of the log fold changes for visualizations or for ranking genes (e.g. GSEA).

```{r lfc_shrink}
lfc_shrink <- lfcShrink(dds, contrast=contrast, type="ashr")

lfc_shrink <- lfc_shrink[order(lfc_shrink$pvalue), ]

```

Below are MA-plots -- a scatter plot of log2 fold changes (on the y-axis) versus the mean of normalized counts (on the x-axis).
Here, we show both the default log fold changes (LFC) and the shrunken log fold changes.


```{r ma_plots, fig.width=5, fig.height=5}
DESeq2::plotMA(res, main="Default LFC")

```

```{r ma_plots_shrunken, fig.width=5, fig.height=5}
DESeq2::plotMA(lfc_shrink, main="Shrunken LFC")

```

## Output results

### Output DE results

Here, we merge the different gene name columns to the DE results and output to a tab-delimited file, which can be opened in Excel for manual perusal. This file is **`r params$comparison_name`_de_res.tsv**, within the *[deseq2_tables/](../extras/deseq2_tables/)*  directory.

```{r out_de_res_prep}
df <- cbind(as.data.frame(rowData(dds)[rownames(lfc_shrink), 1:4]),
            as.data.frame(lfc_shrink)) %>%
  tibble::rownames_to_column("ens_gene")
```

```{r out_de_res}
write_tsv(df, file.path(outdir, "de_res.tsv"))
write_rds(df, file.path(outdir, "de_res.rds"))
```

```{r specific_genes_check}
# determine whether any genes have been specified in the config file
genes_specified <- !(params$genes_of_interest %in% c("", "False", "FALSE", "None"))
if (genes_specified) {
  genes_of_interest <- params$genes_of_interest %>% strsplit(",") %>% unlist() %>% gsub("\\s+", "", .)
}
```

`r if(genes_specified){"### Look for specific genes\nIf we're interested in specific genes (such as genes we'd expect, a priori, to be different),\nwe can check if they are differentially expressed.\nThis is useful for quickly checking if genes of interest are differentially expressed, without having to manually search through the results table.\n\nThe following genes are shown, based on genes specified in the 'genes_of_interest' parameter in the config file."}` 

`r if(genes_specified){"The genes of interest are: "}`
`r if(genes_specified){paste(genes_of_interest, collapse=", ")}`


```{r specific_genes, eval=genes_specified, echo=genes_specified, include=genes_specified}
# message("Genes of interest: ", paste(genes_of_interest, collapse=", "))
df %>% 
  dplyr::filter(Symbol %in% genes_of_interest) %>% 
  dplyr::mutate(log2FoldChange = format(log2FoldChange, scientific=FALSE, digits=3),
                lfcSE = format(lfcSE, scientific=FALSE, digits=3),
                pvalue = format(pvalue, scientific=TRUE, digits=3),
                padj = format(padj, scientific=TRUE, digits=3)) %>%
  knitr::kable() %>%
  kableExtra::kable_styling() %>%
  kableExtra::column_spec(9:10, width = "6em")
```

### Output tables with raw counts

Some folks also find it useful to have tables of the raw counts or the normalized counts. The raw counts can be extracted from the DESeq2 object using either `assay()` or `counts()`. This file is **`r params$comparison_name`_counts.tsv**, within the *[deseq2_tables/](../extras/deseq2_tables/)*  directory.

```{r out_counts_prep}
df <- cbind(as.data.frame(rowData(dds)[, 1:4]),
            assay(dds, "counts")) %>%
  tibble::rownames_to_column("ens_gene")
```

```{r out_counts}
write_tsv(df, file.path(outdir, "counts.tsv"))
```

### Output tables with log2 normalized counts

For the log2 normalized counts, we commonly use the variance stabilized transformation ([VST](https://rdrr.io/bioc/DESeq2/man/varianceStabilizingTransformation.html)). These values can be used for heatmaps, clustering or other downstream applications.
This file is **`r params$comparison_name`_vst.tsv**, within the *[deseq2_tables/](../extras/deseq2_tables/)*  directory.

```{r out_vst_prep}
vsd <- vst(dds, blind=FALSE)

vst_df <- as.data.frame(cbind(rowData(vsd)[, 1:4], assay(vsd))) %>%
  tibble::rownames_to_column("ens_gene")
```

```{r out_vst}
write_rds(vsd, file.path(outdir, "vsd.rds"))
write_tsv(vst_df, file.path(outdir, "vst.tsv"))

```

## Some common plots for DEG analysis

The plots below are some common visualizations for differential expression analysis. Individual plots can be found in the *[deseq2_figures/`r params$comparison_name`/](../extras/deseq2_figures/`r params$comparison_name`/)* directory, for easy access to pdf and png files.

### PCA

Principal Component Analysis (PCA) is a dimensionality reduction technique that 
transforms data into a new coordinate system, where the axes (principal components, or PCs)
are ordered by the amount of variance they capture (shown as percentages in the plots below).
PC 1 captures the most variance in the data, followed by PC 2, and so on.
PCA is often used to visualize the variance in data and to identify how samples cluster together based on their expression profiles.

If group has a significant effect on the variance in the data, we will see that samples cluster together by group in the PCA plot.
If we see samples clustering by some *other* variable (e.g. batch, treatment, age, sex, genotype, etc.), this suggests that this other variable is confounding the results and should be accounted for in the analysis. 
If this is the case, you may need to include this variable as a covariate in the model design (e.g. `design = ~ batch + group`).

**Note, again, that no covariates are included in this analysis**; if you have a more complex 
experimental setup with additional covariates and/or confounders, 
the model used in this report is overly simplistic and won't accurately model your data --
please modify and/or contact the BBC.

Here, we will plot the first four principal components, colored by group.

```{r pca_func}
# Make PCA plots
make_PCA <- function(in_dds, cols, PCx=1, PCy=2, ntop){
  if (ntop > nrow(in_dds)) {
    ntop <- nrow(in_dds)
  }
  mat <- assay(in_dds) 
  col_data <- as.data.frame(colData(in_dds))
  
  # row variances
  vars <- genefilter::rowVars(mat)
  vars_rank <- order(vars, decreasing = TRUE)
  
  # get most variable genes
  mat <- mat[vars_rank[1:ntop], ]
  
  pcaData <- prcomp(t(mat))
  
  prop_var <- data.frame(t(summary(pcaData)$importance))
  names(prop_var) = c("sd", "prop", "cum")
  prop_var$num = 1:nrow(prop_var)
  
  #pcaData <- plotPCA(in_dds, intgroup=colnames(colData(in_dds)), returnData=TRUE, ...)
  percentVar <- round(100 * prop_var$prop)
  
  
  df <- cbind(pcaData$x, col_data)
  
  out_patchwork <- list()
  
  for (i in 1:length(cols)){
    gg_args <- list(x=str_glue("PC{PCx}"), y=str_glue("PC{PCy}"), color=cols[i], label="sample")
    gg_args <- lapply(gg_args, function(x) if (!is.null(x)) sym(x))
    out_patchwork <- c(out_patchwork, list(ggplot(df, aes(!!!gg_args)) +
                                             geom_point(size=1) +
                                             scale_color_manual(values = setNames(c("#440154FF","#2A788EFF", ggsci::pal_npg()(length(levels(dds$group))-2)),
                                                                                  levels(dds$group))) +
                                             xlab(paste0(str_glue("PC{PCx}: "),percentVar[PCx],"% variance")) +
                                             ylab(paste0(str_glue("PC{PCy}: "),percentVar[PCy],"% variance")) + 
                                             theme_bw() + 
                                             ggrepel::geom_text_repel(max.overlaps=5)))
  }
  wrap_plots(out_patchwork)
}
```

```{r plot_pca, fig.height=4.5, fig.width=9}
(make_PCA(vsd, ntop=10000, cols="group") | 
    (make_PCA(vsd, ntop=10000, cols="group", PCx = 3, PCy = 4))) +
  plot_layout(guides="collect") + 
  plot_annotation(title = "PCA - All Samples") 
```

```{r test_cluster_separation, fig.height=4.5, fig.width=9}
test_clustering <- function(in_dds, ntop) {
  if (ntop > nrow(in_dds)) {
    ntop <- nrow(in_dds)
  }

  mat <- assay(in_dds) 
  col_data <- as.data.frame(colData(in_dds))
    
  # row variances
  vars <- genefilter::rowVars(mat)
  vars_rank <- order(vars, decreasing = TRUE)

  # get most variable genes
  mat <- mat[vars_rank[1:ntop], ]

  mat_PERMANOVA <- mat %>%
    t() %>%
    scale() 
  mat_PERMANOVA_df <- mat_PERMANOVA %>%
    as.data.frame() %>%
    tibble::rownames_to_column("sample") %>%
    left_join(., as.data.frame(colData(vsd)), by="sample") 

  adonis2(mat_PERMANOVA ~ group, data = mat_PERMANOVA_df, method='eu')
}
res_clustering <- test_clustering(vsd, ntop=10000)
res_clustering
```

The PERMANOVA test above tests whether the groups are significantly different from each other based on the most variable genes in the dataset.
A significant p-value (e.g. < 0.05) indicates that the groups are significantly different from each other, and that the PCA plot is likely to show separation between the groups.
Here, we see that the groups appear `r if(res_clustering[[1,"Pr(>F)"]] >= 0.05) "*not*"` to be  significantly different from each other 
(p-value = `r format.pval(res_clustering[[1,"Pr(>F)"]], digits=3)`), indicating that the PCA plot is `r if(res_clustering[[1,"Pr(>F)"]] >= 0.05) "not"` likely to show separation between the groups.

### Volcano plot

Volcano plots are a common way to visualize the results of differential expression analyses.
They plot a fold change metric (here, log2FoldChange) on the x-axis and a significance metric on the y-axis (here, the -log10 p-value). 
Genes that are significantly differentially expressed will appear as points in the upper right or upper left areas of the plot, depending on whether they are upregulated or downregulated, respectively.
The directionality of the plot is dependent on which group was specified as the
`group_test` and `group_reference` in the comparisons.tsv -- genes that are 
higher in your test group (here, "`r params$group_test`") will show up on the right side of the plot (upregulated).
Genes that are higher in your reference group (here, "`r params$group_reference`") will show up on the left side of the plot (downregulated).

Below is a volcano plot, with labels for the top 10 genes ranked by absolute log fold change.


```{r make_volc_func}
make_volcano <- function(df, pval_nm, pval_cutoff=0.1){
  # remove genes with NA for pvalue
  df <- df[which(!is.na(df[[pval_nm]])), ]

  # add gene names
  df <- cbind(df, rowData(dds)[rownames(df), 1:4])

  top_genes <- df %>%
    dplyr::arrange(desc(abs(df$log2FoldChange))) %>%
    dplyr::filter(row_number() <= 10) %>%
    rownames()

  df$Sig <- ifelse(df$padj <= pval_cutoff, "Sig", "NS")

  df[[pval_nm]] <- -log10(df[[pval_nm]])

  ggplot(df, aes(x=log2FoldChange, y=.data[[pval_nm]])) +
    geom_point(aes(color=Sig), size=0.6) +
    scale_color_manual(values=c("black", "salmon")) +
    theme_bw() + ylab(str_glue("-log10(", pval_nm,")")) +
    geom_text_repel(data=df[top_genes, ],
                    aes(label=Uniq_syms), max.overlaps=Inf, min.segment.length = 0)
}
```

```{r volcano, fig.width=4, fig.height=4}
make_volcano(as.data.frame(lfc_shrink),
             pval_nm="padj", pval_cutoff=fdr_cutoff)
```

### Heatmap

Heatmaps are a common way to visualize the expression of genes across samples.
Here, the top 20 genes (by absolute log2FoldChange) are selected and plotted,
colored by group.

```{r heatmap, fig.width=6, fig.height=6}
top_genes <- rownames(res)[1:20]

top_se <- se[top_genes, ]
mat <- assay(top_se, "vst")
mat <- t(scale(t(mat), scale=FALSE, center = TRUE))

# column annot
ht_col_annot <- as.data.frame(colData(top_se)[, "group", drop=FALSE])

group_lvls <- unique(ht_col_annot$group)
ht_col_colors <- list(group=setNames(c("#440154FF","#2A788EFF", ggsci::pal_npg()(length(group_lvls)-2)),
                                         nm=group_lvls))

Heatmap(mat,
        name = "Mean-centered",
        cluster_columns = FALSE,
        row_labels=rowData(top_se)$Uniq_syms,
        show_column_names = FALSE,
        top_annotation=HeatmapAnnotation(df=ht_col_annot,
                                         col=ht_col_colors),
        column_title = "Top DE genes",
        row_title = paste0(nrow(mat), " genes")
)


```

### P value distribution

The distribution of p-values can give us an idea of the number of differentially expressed genes (DE genes) in our analysis,
and whether the model used for differential expression is appropriate.
We expect the p-values to be uniformly distributed if there are not many DE genes, or anti-conservative (more p-values close to 0) if there are many DE genes.
See [here](http://varianceexplained.org/statistics/interpreting-pvalue-histogram/) for more details about how to interpret these,
and again, contact the BBC if you have questions about your results.

```{r pval, fig.width=4, fig.height=4}

ggplot(data = as.data.frame(lfc_shrink) %>%
         dplyr::filter(!is.na(pvalue)),
       aes(x = pvalue)) +
  geom_histogram(color = "black", fill = "gray55",
                 breaks = seq(0, 1, 0.05)) + theme_bw() + theme(plot.title=element_text(size=10))

```

# Appendix

*Want to make sense of your differential expression results on a broader level?*

See [gsea_`r params$comparison_name`.html](gsea_`r params$comparison_name`.html) for a pathway-level analysis using GSEA (Gene Set Enrichment Analysis), 
a method for determining whether the changes you see between your groups (`r params$group_test` and `r params$group_reference`) are similar to known pathways/genesets. The report contains a summary of the results, including the top enriched pathways, as well as visualizations of the results. This can help you understand the biological significance of your results and identify potential pathways that are affected by the changes in gene expression.

And, as always, contact the BBC if you have questions about your results! 


## SessionInfo

```{r sessioninfo}
sessionInfo()
```

## Runtime

```{r endtime}
# output time taken to run script
end_tm <- Sys.time()
# end_tm
# end_tm - start_tm

```

This analysis was completed on `r format(end_tm, "%B %d, %Y at %H:%M")`, with a total runtime of
 `r ifelse(difftime(end_tm, start_tm, units = "mins") < 1, paste0(round(difftime(end_tm, start_tm, units = "secs"), 2), " seconds"), paste0(round(difftime(end_tm, start_tm, units = "mins"), 2), " minutes"))`.
